targets/pna_nic/pna_nic.cpp

```diff
@@ -0,0 +1,235 @@
+/* Copyright 2024 Marvell Technology, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Rupesh Chiluka (rchiluka@marvell.com)
+ *
+ */
+
+#include <bm/bm_sim/parser.h>
+#include <bm/bm_sim/tables.h>
+#include <bm/bm_sim/logger.h>
+
+#include <unistd.h>
+
+#include <iostream>
+#include <fstream>
+#include <string>
+
+#include "pna_nic.h"
+
+namespace {
+
+struct hash_ex {
+  uint32_t operator()(const char *buf, size_t s) const {
+    const uint32_t p = 16777619;
+    uint32_t hash = 2166136261;
+
+    for (size_t i = 0; i < s; i++)
+      hash = (hash ^ buf[i]) * p;
+
+    hash += hash << 13;
+    hash ^= hash >> 7;
+    hash += hash << 3;
+    hash ^= hash >> 17;
+    hash += hash << 5;
+    return static_cast<uint32_t>(hash);
+  }
+};
+
+struct bmv2_hash {
+  uint64_t operator()(const char *buf, size_t s) const {
+    return bm::hash::xxh64(buf, s);
+  }
+};
+
+}  // namespace
+
+// if REGISTER_HASH calls placed in the anonymous namespace, some compiler can
+// give an unused variable warning
+REGISTER_HASH(hash_ex);
+REGISTER_HASH(bmv2_hash);
+
+extern int import_primitives();
+
+namespace bm {
+
+namespace pna {
+
+packet_id_t PnaNic::packet_id = 0;
+
+
+PnaNic::PnaNic(bool enable_swap)
+  : Switch(enable_swap),
+    input_buffer(1024),
+    output_buffer(128),
+    // cannot use std::bind because of a clang bug
+    // https://stackoverflow.com/questions/32030141/is-this-incorrect-use-of-stdbind-or-a-compiler-bug
+    my_transmit_fn([this](port_t port_num, packet_id_t pkt_id,
+                          const char *buffer, int len) {
+        _BM_UNUSED(pkt_id);
+        this->transmit_fn(port_num, buffer, len);
+    }),
+    pre(new McSimplePreLAG()),
+    start(clock::now())
+    {
+  add_component<McSimplePreLAG>(pre);
```

Comment:
does the `McSimplePreLAG` (multicast engine with LAG support) makes sense for PNA? If not, or if we plan on having a different multicast implementation, it should be removed from this PR and not added as a component.

Reply:
The PNA specification still needs to decide where multicast replication occurs: https://p4.org/p4-spec/docs/PNA-v0.7.html#packet-path-details.

Wiil remove `McSimplePreLAG` for now.